cmd/titus-storage/ebs.go

```diff
@@ -0,0 +1,312 @@
+	for {
+		if ctr >= len(runes) {
+			return "", fmt.Errorf("Ran out of drive names")
+		}
+		if !doesDriveExist(deviceName + string(runes[ctr])) {
+			break
+		}
+		ctr++
+	}
+	return deviceName + string(runes[ctr]), nil
+}
+
+func doesDriveExist(driveName string) bool {
+	if _, err := os.Stat(driveName); os.IsNotExist(err) {
+		for _, file := range listNvmeBlockDevices() {
+			if d, _ := ebsnvme.ScanDevice(file); d.Name == driveName {
+				return true
+			}
+		}
+		return false
+	}
+	return true
+}
+
+// waitForDeviceToDisappear is an inexpensive way to poll for an EBS
+// detachment without calling the EC2 API. It does this by waiting
+// for the block device it was attached with to dissapear from /dev
+func waitForDeviceToDisappear(ctx context.Context, driveName string, maxAttempts int) error {
+	l := logger.GetLogger(ctx)
+	var attempts int
+	for doesDriveExist(driveName) {
+		time.Sleep(2 * time.Second)
+		attempts++
+		if attempts >= maxAttempts {
+			return fmt.Errorf("waited the max %d seconds for %s to go, but it is still here", attempts*2, driveName)
+		}
+	}
+	l.Printf("waited for %s to disappear, and it is now gone after %d seconds", driveName, attempts*2)
+	return nil
+}
+
+// waitForDriveToExistWithTimeout is an inexpensive way to poll for an EBS
+// attachment without calling the EC2 API. It does this by scaning the
+// list of nvme volumes on disk and looking at their 'serial number'
+func waitForDriveToExistWithTimeout(driveName string, maxAttempts int) error {
+	var attempts int
+	for !doesDriveExist(driveName) {
+		time.Sleep(2 * time.Second)
+		attempts++
+		if attempts >= maxAttempts {
+			return fmt.Errorf("drive %s still does't exist after waiting %d seconds", driveName, attempts*2)
+		}
+	}
+	return nil
+}
+
+// GetActualBlockDeviceName returns the actual name of the block device seen
+// within the instance, like /dev/nvme1
+// This is often different from the ec2 device name, which is usually
+// something like /dev/sdx
+func GetActualBlockDeviceName(name string) (string, error) {
+	for _, device := range listNvmeBlockDevices() {
+		if d, _ := ebsnvme.ScanDevice(device); d.Name == name {
+			return device, nil
+		}
+	}
+	if _, err := os.Stat(name); os.IsNotExist(err) {
+		return "", err
+	}
+	return name, nil
+}
+
+func listNvmeBlockDevices() (devices []string) {
+	re := regexp.MustCompile(`(^\/dev\/nvme[0-9]+n1$)`)
+	f, _ := filepath.Glob("/dev/nvme*")
+	for _, d := range f {
+		if re.Match([]byte(d)) {
+			devices = append(devices, d)
+		}
+	}
+	return
+}
+
+// getDeviceName scans the Nvme on disk and looks at the
+// serial numbers of the devices and returns the device name
+// (/dev/nmve1). Behaves like a hash get, returning key, ok.
+// returning "", false if unfound.
+func getDeviceName(ebsVolumeID string) (string, bool) {
+	deviceName := ""
+	for _, device := range listNvmeBlockDevices() {
+		if d, _ := ebsnvme.ScanDevice(device); d.VolumeID == ebsVolumeID {
+			deviceName = d.Name
```

Comment:
Cant you just return from here? Or can multiple devices have the right volume ID? If so, which one should it be?

Reply:
oh yea of course. No I don't think we could ever have 2 devices with the same volume id